service/src/test/java/bio/terra/workspace/service/resource/controlled/cloud/aws/s3StorageFolder/AwsS3StorageFolderStepTest.java

```diff
@@ -0,0 +1,353 @@
+import software.amazon.awssdk.services.s3.model.ListObjectsV2Request;
+import software.amazon.awssdk.services.s3.model.ListObjectsV2Response;
+import software.amazon.awssdk.services.s3.model.PutObjectRequest;
+import software.amazon.awssdk.services.s3.model.PutObjectResponse;
+import software.amazon.awssdk.services.s3.model.S3Error;
+import software.amazon.awssdk.services.s3.model.S3Exception;
+import software.amazon.awssdk.services.s3.model.S3Object;
+
+@TestInstance(Lifecycle.PER_CLASS)
+public class AwsS3StorageFolderStepTest extends BaseAwsUnitTest {
+
+  @Mock private FlightContext mockFlightContext;
+  @Mock private AwsCloudContextService mockAwsCloudContextService;
+  @Mock private SamService mockSamService;
+  @Mock private S3Client mockS3Client;
+
+  private MockedStatic<AwsUtils> mockAwsUtils;
+  private ControlledAwsS3StorageFolderResource s3FolderResource;
+  private static final AwsServiceException s3Exception1 =
+      S3Exception.builder().message("not authorized to perform").build();
+
+  @BeforeAll
+  public void init() {
+    mockAwsUtils = Mockito.mockStatic(AwsUtils.class);
+    s3FolderResource =
+        ControlledAwsResourceFixtures.makeDefaultAwsS3StorageFolderResource(
+            ControlledResourceFixtures.WORKSPACE_ID);
+  }
+
+  @BeforeEach
+  public void setup() {
+    when(mockFlightContext.getResult())
+        .thenReturn(new StepResult(StepStatus.STEP_RESULT_FAILURE_FATAL));
+
+    when(mockAwsCloudContextService.getAwsCloudContext(any()))
+        .thenReturn(Optional.of(ControlledAwsResourceFixtures.makeAwsCloudContext()));
+
+    when(mockSamService.getSamUser((AuthenticatedUserRequest) any()))
+        .thenReturn(WorkspaceFixtures.SAM_USER);
+
+    mockAwsUtils.clearInvocations();
+    mockAwsUtils
+        .when(() -> AwsUtils.createWsmCredentialProvider(any(), any()))
+        .thenReturn(ControlledAwsResourceFixtures.AWS_CREDENTIALS_PROVIDER);
+    mockAwsUtils.when(() -> AwsUtils.getS3Client(any(), any())).thenReturn(mockS3Client);
+    mockAwsUtils.when(() -> AwsUtils.checkException(any())).thenCallRealMethod();
+  }
+
+  @Test
+  public void createS3FolderTest() throws InterruptedException {
+    CreateAwsS3StorageFolderStep createS3FolderStep =
+        new CreateAwsS3StorageFolderStep(
+            s3FolderResource,
+            mockAwsCloudContextService,
+            MockMvcUtils.USER_REQUEST,
+            mockSamService);
+
+    mockAwsUtils
+        .when(() -> AwsUtils.createStorageFolder(any(), any(), any()))
+        .thenAnswer(invocation -> null) /* success */
+        .thenThrow(WorkspaceFixtures.API_EXCEPTION);
+
+    // do: success
+    assertThat(
+        createS3FolderStep.doStep(mockFlightContext), equalTo(StepResult.getStepResultSuccess()));
+
+    // do: error
+    assertThrows(
+        WorkspaceFixtures.API_EXCEPTION.getClass(),
+        () -> createS3FolderStep.doStep(mockFlightContext));
+
+    // undo: same as tests for DeleteAwsS3StorageFolderStep
```

Comment:
Shouldn't this check that executeDeleteAwsS3StorageFolder is called? Same for createS3FolderTestFull below

Reply:
long response & more details --> 

- The only reason there is a static executeDeleteAwsS3StorageFolder function is to ensure create.undo shares the same code as delete.do (do/undo are instance methods)
- Skipped the undo test in create step since it is the same as delete step (already has tests)
- We could `verify that x is called` check, but didnt add earlier since it would add very little value here: Connected test would cover it

adding it now for the sake of completeness